src/ai/backend/agent/resources.py

```diff
@@ -444,6 +469,303 @@ def get_additional_allowed_syscalls(self) -> list[str]:
+        return reserved_slots
+
+    def _calculate_resource_scaling_factor(self, allocated_slots: SlotsMap) -> SlotsMap:
+        match self.local_config.resource.allocation_mode:
+            case ResourceAllocationMode.SHARED:
+                return defaultdict(lambda: Decimal(1.0))
+            case ResourceAllocationMode.AUTO_SPLIT:
+                return defaultdict(lambda: Decimal(1.0) / Decimal(self.num_agents))
+            case ResourceAllocationMode.MANUAL:
+                total_slots = self.total_slots
+                if SlotName("cpu") not in allocated_slots or SlotName("cpu") not in total_slots:
+                    raise ValueError("CPU not in allocated or total slots seen")
+                if SlotName("mem") not in allocated_slots or SlotName("mem") not in total_slots:
+                    raise ValueError("Memory not in allocated or total slots seen")
+                scaling_factor = {
+                    slot_name: slot / self.available_total_slots[slot_name]
+                    for slot_name, slot in allocated_slots.items()
+                }
+                return scaling_factor
+
+    async def _load_resources(self) -> Mapping[DeviceName, AbstractComputePlugin]:
+        local_config_dump = self.local_config.model_dump(by_alias=True)
+
+        match self.local_config.agent_common.backend:
+            case AgentBackend.DOCKER:
+                from .docker.resources import load_resources as docker_load
+
+                return await docker_load(self.etcd, local_config_dump)
+            case AgentBackend.KUBERNETES:
+                from .kubernetes.resources import load_resources as kubernetes_load
+
+                return await kubernetes_load(self.etcd, local_config_dump)
+            case AgentBackend.DUMMY:
+                from .dummy.config import DEFAULT_CONFIG_PATH, dummy_local_config
+                from .dummy.resources import load_resources as dummy_load
+
+                raw_config, _ = read_from_file(DEFAULT_CONFIG_PATH, "dummy")
+                dummy_config = dummy_local_config.check(raw_config)
+                return await dummy_load(self.etcd, local_config_dump, dummy_config)
+
+    async def _scan_available_resources(self) -> Mapping[SlotName, Decimal]:
+        compute_device_types = {name: cctx.instance for name, cctx in self.computers.items()}
+
+        match self.local_config.agent_common.backend:
+            case AgentBackend.DOCKER:
+                from .docker.resources import scan_available_resources as docker_scan
+
+                return await docker_scan(compute_device_types)
+            case AgentBackend.KUBERNETES:
+                from .kubernetes.resources import scan_available_resources as kubernetes_scan
+
+                return await kubernetes_scan(compute_device_types)
+            case AgentBackend.DUMMY:
+                from .dummy.resources import scan_available_resources as dummy_scan
+
+                return await dummy_scan(compute_device_types)
```

Comment:
This change, in terms of extensibility, rather seems like a regressive implementation and doesn't look good.